### PR TITLE
Add more information to outputs (add instance indices for error types)

### PR DESCRIPTION
**Pull request by jackboyla**
Hi 👋 I would like `nervaluate`'s results to show me the examples where my NER model makes mistakes, i.e false positive/false negatives. This functionality has already been mentioned in issue #68.

I have made some simple additions to `evaluate.py`, but this will break a lot of tests. I'm happy to adapt them, but I want to first check if we are happy with this form of additional information before I continue 😄 

Here's an example of how the changes affect the output:

```
from nervaluate import Evaluator


def test_evaluator_simple_case():
    true = [
        [{"label": "PER", "start": 2, "end": 4}],
        [
            {"label": "LOC", "start": 1, "end": 2},
        ],
        [
            {"label": "LOC", "start": 1, "end": 2},
            {"label": "LOC", "start": 3, "end": 4}, # missed
        ],
        [
            {"label": "PER", "start": 27, "end": 29},
        ],
        [
            {"label": "LOC", "start": 4, "end": 7},
        ],
    ]
    pred = [
        [{"label": "PER", "start": 2, "end": 4}],
        [
            {"label": "LOC", "start": 1, "end": 2},
            {"label": "PER", "start": 13, "end": 14}, # false positive (spurious)
        ],
        [
            {"label": "LOC", "start": 1, "end": 2},
        ],
        [
            {"label": "PER", "start": 28, "end": 31}, # partial
        ],
        [
            {"label": "LOC", "start": 4, "end": 7},
            {"label": "LOC", "start": 24, "end": 26}, # another false positive (spurious)
        ],
    ]
    evaluator = Evaluator(true, pred, tags=["LOC", "PER"])
    results, results_agg = evaluator.evaluate()

    return results, results_agg
```

Our overall results will look like:

```
{'ent_type': {'correct': 5,
  'incorrect': 0,
  'incorrect_indices': [],
  'partial': 0,
  'partial_indices': [],
  'missed': 1,
  'missed_indices': [2],
  'spurious': 2,
  'spurious_indices': [1, 4],
  'possible': 6,
  'actual': 7,
  'precision': 0.7142857142857143,
  'recall': 0.8333333333333334,
  'f1': 0.7692307692307692},
 'partial': {'correct': 4,
  'incorrect': 0,
  'incorrect_indices': [],
  'partial': 1,
  'partial_indices': [3],
  'missed': 1,
  'missed_indices': [2],
  'spurious': 2,
  'spurious_indices': [1, 4],
  'possible': 6,
  'actual': 7,
  'precision': 0.6428571428571429,
  'recall': 0.75,
  'f1': 0.6923076923076924},
 'strict': {'correct': 4,
  'incorrect': 1,
  'incorrect_indices': [3],
  'partial': 0,
  'partial_indices': [],
  'missed': 1,
  'missed_indices': [2],
  'spurious': 2,
  'spurious_indices': [1, 4],
  'possible': 6,
  'actual': 7,
  'precision': 0.5714285714285714,
  'recall': 0.6666666666666666,
  'f1': 0.6153846153846153},
 'exact': {'correct': 4,
  'incorrect': 1,
  'incorrect_indices': [3],
  'partial': 0,
  'partial_indices': [],
  'missed': 1,
  'missed_indices': [2],
  'spurious': 2,
  'spurious_indices': [1, 4],
  'possible': 6,
  'actual': 7,
  'precision': 0.5714285714285714,
  'recall': 0.6666666666666666,
  'f1': 0.6153846153846153}}
```

The relevant indices will also be added on the per-tag output.

### CODE REVIEW

**Comment by davidsbatista** [comment]
Hi @jackboyla, and thanks a lot for you collaboration.

Please don't use `self.results` to store the indices, define another structure. I believe this should be done in a different structure to keep things clean and separate; this might also reduce the number of failed tests. Maybe you can follow the same philosophy as in `self.results`, i.e, define a base dictionary and then use `deepcopy`.

Going back to the original issue:

``Is there a way to find out for which instance during evaluation was marked under 'correct' or 'incorrect' or 'spurious', etc for a particular evaluation schema?``

 I see that the question was to have a way to find which instances were marked under correct, incorrect, spurious, etc.

Maybe it's also useful to have a function that prints a nice output to the console with those entities, just as indices, but optionally if the text is given text surface strings themselves.

What do you think @ivyleavedtoadflax ?

By the way, late happy New Year to you all :)

File: src/nervaluate/evaluate.py [hunk 1]
```diff
@@ -60,16 +79,22 @@ def evaluate(self) -> Tuple[Dict, Dict]:
         if len(self.true) != len(self.pred):
             raise ValueError("Number of predicted documents does not equal true")
 
-        for true_ents, pred_ents in zip(self.true, self.pred):
+        for index, (true_ents, pred_ents) in enumerate(zip(self.true, self.pred)):
             # Compute results for one message
-            tmp_results, tmp_agg_results = compute_metrics(true_ents, pred_ents, self.tags)
+            tmp_results, tmp_agg_results, tmp_results_indices, tmp_agg_results_indices = compute_metrics(true_ents, pred_ents, self.tags, index)
+
+            # import ipdb; ipdb.set_trace()
```
Comment (davidsbatista) [inline src/nervaluate/evaluate.py:86]: leftovers

File: src/nervaluate/evaluate.py [hunk 2]
```diff
@@ -81,17 +106,21 @@ def evaluate(self) -> Tuple[Dict, Dict]:
...
+        return self.results, self.evaluation_agg_entities_type, self.evaluation_indices, self.evaluation_agg_indices
 
 
 # flake8: noqa: C901
 def compute_metrics(  # type: ignore
-    true_named_entities, pred_named_entities, tags: List[str]
+    true_named_entities, pred_named_entities, tags: List[str], index: int = 0
 ):  # pylint: disable=too-many-locals, too-many-branches, too-many-statements
     """
     Compute metrics on the collected true and predicted named entities
```
[the remark this answers is not in the document]
Comment (davidsbatista) [reply src/nervaluate/evaluate.py:126]: Also, the variable name `index` it's a bit vague, maybe renaming it to `pred_ent_index` or `just ent_index` ?

File: src/nervaluate/evaluate.py [hunk 3]
```diff
@@ -128,6 +157,23 @@ def compute_metrics(  # type: ignore
     # results by entity type
     evaluation_agg_entities_type = {e: deepcopy(evaluation) for e in tags}
 
+    eval_indices = {
```
Comment (davidsbatista) [inline src/nervaluate/evaluate.py:160]: maybe `eval_entities` - it's more clear one read the codes

Comment (davidsbatista) [inline src/nervaluate/evaluate.py:160]: rename to `eval_ent_indices`

File: src/nervaluate/evaluate.py [hunk 4]
```diff
@@ -128,6 +157,23 @@ def compute_metrics(  # type: ignore
...
+        "correct_indices": [],
+        "incorrect_indices": [],
+        "partial_indices": [],
+        "missed_indices": [],
+        "spurious_indices": [],
+    }
+
+    # Create dicts to hold indices for correct/spurious/missing/etc examples
+    evaluation_indices = {
```
Comment (davidsbatista) [inline src/nervaluate/evaluate.py:169]: rename to rename to `evaluation_ent_indices`

File: src/nervaluate/evaluate.py [hunk 5]
```diff
@@ -128,6 +157,23 @@ def compute_metrics(  # type: ignore
...
+        "strict": deepcopy(eval_indices),
+        "ent_type": deepcopy(eval_indices),
+        "partial": deepcopy(eval_indices),
+        "exact": deepcopy(eval_indices),
+    }
+    evaluation_agg_indices = {e: deepcopy(evaluation_indices) for e in tags}
```
Comment (davidsbatista) [inline src/nervaluate/evaluate.py:175]: rename to `evaluation_agg_ent_indices`

**Comment by davidsbatista** [comment]
Hi @jackboyla, thanks for your updates - I've left a few comments, mostly regarding variable names. I fear that this code starts to become a bit spaghetti; that's the reason I'm so strict.

By the way, did you run the code quality shakes locally?

**Comment by davidsbatista** [comment]
> You can see for 'correct_indices': [0, 1, 1, 2, 4], the instance at index 1 had two correct predictions (hence 1 appears twice), but it will not specify which predictions within the instance were correct.

Sorry, I am a bit confused about how this can happen? How can a predicted entity be correct twice under the same evaluation scenario. Maybe you can show me an example? As mentioned earlier, I haven't looked at and touched the code in a while.

I thought that the index represents each instance in the prediction list, as shown below:

```
pred = [
        [{"label": "PER", "start": 2, "end": 4}],
        [
            {"label": "LOC", "start": 1, "end": 2},
            {"label": "PER", "start": 13, "end": 14}, # false positive (spurious)
        ],
        [
            {"label": "LOC", "start": 1, "end": 2},
        ],
        [
            {"label": "PER", "start": 28, "end": 31}, # partial
        ],
        [
            {"label": "LOC", "start": 4, "end": 7},
            {"label": "LOC", "start": 24, "end": 26}, # another false positive (spurious)
        ],
    ]
```

`I can add this information so each element is a tuple (instance_index, entity_within_instance_index), but I'm afraid this will make the output too verbose:
`

Regarding this suggestion, I think this starts to become really convoluted. The end user just wants a nice report, in this case, the entities themselves or the offsets of the entities in the document which were incorrect under each scenario.

Please, see the two functions of the below:

- summary_report_ent()
- summary_report_overall()

If you run those functions, you will see a nice report, see the folder examples, to run in example that cause this functions.

We can use the format that you proposed as some intermediate representation, but I don't think that should ever be exposed to the user. You can rely on that intermediate representation to output a nice report to the console or to a file.

Sorry if this seems more work than you had initially thought, but I think it's better if we tackle this in a structured and clean way to avoid having more and more spaghetti code.

**Comment by jackboyla** [comment]
Thanks for taking the time to give this feedback 😄 I think I didn't make it clear in my last comment. Here's an example:

We have predictions for 3 separate instances:
```
pred = [
        [{"label": "PER", "start": 2, "end": 4}], # correct
        [
            {"label": "LOC", "start": 1, "end": 2}, # correct
            {"label": "LOC", "start": 4, "end": 5}, # false positive (spurious)
            {"label": "PER", "start": 13, "end": 14}, # false positive (spurious)
        ],
       [{"label": "PER", "start": 7, "end": 9}], # false positive (spurious)
    ]
```

We see that for instance at index `1`, there are two `spurious` predictions for two separate entities -- at indices `1` and `2` **inside** that instance. So the current implementation records 2 errors under spurious. Additionally, instance `2` contains one prediction which is spurious:

```
{'strict': {...,
  ...,
  'spurious_indices': [1, 1, 2]},
...
}
```

This tells the user that 2 `spurious` errors have been recorded in instance `1`. Alternatively, we can just return a `set()` of these indices, but then the user doesn't know how many `spurious` errors are present in instance `1`.

I then proposed:

> I can add this information so each element is a tuple (instance_index, entity_within_instance_index), but I'm afraid this will make the output too verbose: 

This would include the position of the entities __within__ the instance where the error occurred:

```
{'strict': {...,
  ...,
  'spurious_indices': [(1 1), (1, 2), (2,0)]},
...
}
```

With regards to the above, I do agree that it becomes too convoluted, and it would be impossible to print this nicely when dealing with many instances and entities.

Looking at it from the end-user's point of view, they just want an output that shows them at what instances the NER model failed. If there is an error in the instance, the user should look at the entire example.

Regarding the printing, I can do it in the style of `summary_report_ent` and `summary_report_overall`, but it may not be very pretty: if there are many instances, the output will not fit on screen.

I hope this makes it a bit clearer what I'm trying to say 😃

**Comment by jackboyla** [comment]
Hi @davidsbatista, I've added print functions for both overall evaluation indices and per-entity indices -- `summary_report_overall_indices` and `summary_report_ents_indices` respectively. They differ from the other summary functions in that they will only print the results from one given evaluation schema (`exact`, `ent_type`, etc..) as printing all schemas is very verbose.

Here's an example:

```
evaluation_indices = {'strict': {'correct_indices': [(0, 0), (1, 0), (1, 1), (2, 0), (4, 0)],
  'incorrect_indices': [(3, 0)],
  'partial_indices': [],
  'missed_indices': [(2, 0)],
  'spurious_indices': [(1, 2), (4, 1)]},
 'ent_type': {'correct_indices': [(0, 0), (1, 0), (1, 1), (2, 0), (4, 0)],
  'incorrect_indices': [],
  'partial_indices': [],
  'missed_indices': [(2, 0)],
  'spurious_indices': [(1, 2), (4, 1)]},
 'partial': {'correct_indices': [(0, 0), (1, 0), (1, 1), (2, 0), (4, 0)],
  'incorrect_indices': [],
  'partial_indices': [(3, 0)],
  'missed_indices': [(2, 0)],
  'spurious_indices': [(1, 2), (4, 1)]},
 'exact': {'correct_indices': [(0, 0), (1, 0), (1, 1), (2, 0), (4, 0)],
  'incorrect_indices': [(3, 0)],
  'partial_indices': [],
  'missed_indices': [(2, 0)],
  'spurious_indices': [(1, 2), (4, 1)]}}
```

You have the option to provide `preds` as an argument:

```
print(summary_report_overall_indices(evaluation_indices, 'partial', preds))
```

which will return:

```
Indices for error schema 'partial':

Correct indices indices:
  - Instance 0, Entity 0: Label=PER, Start=2, End=4
  - Instance 1, Entity 0: Label=LOC, Start=1, End=2
  - Instance 1, Entity 1: Label=LOC, Start=5, End=6
  - Instance 2, Entity 0: Label=LOC, Start=1, End=2
  - Instance 4, Entity 0: Label=LOC, Start=4, End=7

Incorrect indices indices:
  - None

Partial indices indices:
  - Instance 3, Entity 0: Label=PER, Start=28, End=31

Missed indices indices:
  - Instance 2, Entity 0: Label=LOC, Start=1, End=2

Spurious indices indices:
  - Instance 1, Entity 2: Label=PER, Start=13, End=14
  - Instance 4, Entity 1: Label=LOC, Start=24, End=26

```

or do not add `preds`, in which case the function returns:

```
Indices for error schema 'partial':

Correct indices indices:
  - Instance 0, Entity 0
  - Instance 1, Entity 0
  - Instance 1, Entity 1
  - Instance 2, Entity 0
  - Instance 4, Entity 0

Incorrect indices indices:
  - None

Partial indices indices:
  - Instance 3, Entity 0

Missed indices indices:
  - Instance 2, Entity 0

Spurious indices indices:
  - Instance 1, Entity 2
  - Instance 4, Entity 1

```

On an per-entity level, we can use:

```
print(summary_report_ents_indices(evaluation_agg_indices, 'partial', preds))
```

Again, the `preds` is optional. This will return:

```

Entity Type: LOC
  Error Schema: 'partial'
    (LOC) Correct indices:
      - Instance 1, Entity 0: Label=LOC, Start=1, End=2
      - Instance 1, Entity 1: Label=LOC, Start=5, End=6
      - Instance 2, Entity 0: Label=LOC, Start=1, End=2
      - Instance 4, Entity 0: Label=LOC, Start=4, End=7
    (LOC) Incorrect indices:
      - None
    (LOC) Partial indices:
      - None
    (LOC) Missed indices:
      - Instance 2, Entity 0: Label=LOC, Start=1, End=2
    (LOC) Spurious indices:
      - Instance 4, Entity 1: Label=LOC, Start=24, End=26

Entity Type: PER
  Error Schema: 'partial'
    (PER) Correct indices:
      - Instance 0, Entity 0: Label=PER, Start=2, End=4
    (PER) Incorrect indices:
      - None
    (PER) Partial indices:
      - Instance 3, Entity 0: Label=PER, Start=28, End=31
    (PER) Missed indices:
      - None
    (PER) Spurious indices:
      - Instance 1, Entity 2: Label=PER, Start=13, End=14

```

**Comment by davidsbatista** [comment]
Hello @jackboyla and thanks once again for your efforts.

This seems to be in line with what I envision as a report summary. I will approve the workflow so that the call quality checks can be run and we can start reviewing the code together.

**Comment by davidsbatista** [comment]
@ivyleavedtoadflax it seems the coverage badge is giving issues again. Do you have any idea why? Maybe we can disable it or try to find a replacement, why do you think?

**Comment by davidsbatista** [comment]
@ivyleavedtoadflax shall I just open a PR to disable the coverage badge?

**Comment by davidsbatista** [comment]
@ivyleavedtoadflax bump

**Comment by ivyleavedtoadflax** [comment]
Hey sorry for the slow reply. Let's go with the easy option of disabling for now to unblock this PR. Are you ok to put in a PR? 🙏

**Comment by davidsbatista** [comment]
@jackboyla merge the `main` branch into your - I've removed the workflow that was causing this issue

**Comment by davidsbatista** [comment]
done! :) @ivyleavedtoadflax do you know if I have permissions to make a new release?

**Comment by jackboyla** [comment]
great stuff thank you! 😄 I just realised I forgot to include in the README how this changes things. Evaluator.evaluate() now returns 4 variables instead of 2:

```
from nervaluate import Evaluator
true = [
    [{"label": "PER", "start": 2, "end": 4}],
    [{"label": "LOC", "start": 1, "end": 2},
     {"label": "LOC", "start": 3, "end": 4}]
]

pred = [
    [{"label": "PER", "start": 2, "end": 4}],
    [{"label": "LOC", "start": 1, "end": 2},
     {"label": "LOC", "start": 3, "end": 4},
     {"label": "LOC", "start": 12, "end": 14}]
]

evaluator = Evaluator(true, pred, tags=['LOC', 'PER'])

results, results_by_tag, result_indices, result_indices_by_tag = evaluator.evaluate()
```

Additionally, I wanted to include how users can pretty print this new info:

```
from nervaluate import summary_report_ents_indices
print(summary_report_ents_indices(result_indices_by_tag, error_schema='partial', preds=pred))
```

```

Entity Type: LOC
  Error Schema: 'partial'
    (LOC) Correct indices:
      - Instance 1, Entity 0: Label=LOC, Start=1, End=2
      - Instance 1, Entity 1: Label=LOC, Start=3, End=4
    (LOC) Incorrect indices:
      - None
    (LOC) Partial indices:
      - None
    (LOC) Missed indices:
      - None
    (LOC) Spurious indices:
      - Instance 1, Entity 2: Label=LOC, Start=12, End=14

Entity Type: PER
  Error Schema: 'partial'
    (PER) Correct indices:
      - Instance 0, Entity 0: Label=PER, Start=2, End=4
    (PER) Incorrect indices:
      - None
    (PER) Partial indices:
      - None
    (PER) Missed indices:
      - None
    (PER) Spurious indices:
      - None
```

**Comment by davidsbatista** [comment]
feel free open another PR

**Comment by jackboyla** [comment]
ok cool https://github.com/MantisAI/nervaluate/pull/74

**Pull request merged**